sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,364 @@
+
+  return kDifRvTimerOk;
+}
+
+// Computes the appropriate register for a particular interrupt register, given
+// a number of comparators.
+//
+// Currently, all IRQ registers are placed after the comparator registers,
+// so the register offsets given by HW are not useable. The offsets need to be
+// compensated for by adding a double word for each comparator after the first
+// one.
+//
+// We also do not handle the case when comparator_count > 32, which would cause
+// multiple interrupt registers to be generated.
+static ptrdiff_t irq_reg_for_hart(uint32_t hart_id, uint32_t comparators,
+                                  ptrdiff_t reg_offset) {
+  ptrdiff_t base = reg_for_hart(hart_id, reg_offset);
+  ptrdiff_t extra_comparator_offset = sizeof(uint64_t) * (comparators - 1);
+  return base + extra_comparator_offset;
+}
+
+dif_rv_timer_result_t dif_rv_timer_irq_enable(const dif_rv_timer_t *timer,
+                                              uint32_t hart_id,
+                                              uint32_t comp_id,
+                                              dif_rv_timer_enabled_t state) {
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_ENABLE0_REG_OFFSET);
+
+  switch (state) {
+    case kDifRvTimerEnabled:
+      mmio_region_nonatomic_set_bit32(timer->base_addr, reg, comp_id);
+      break;
+    case kDifRvTimerDisabled:
+      mmio_region_nonatomic_clear_bit32(timer->base_addr, reg, comp_id);
+      break;
+    default:
+      return kDifRvTimerBadArg;
+  }
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_irq_get(const dif_rv_timer_t *timer,
+                                           uint32_t hart_id, uint32_t comp_id,
+                                           bool *flag_out) {
+  if (timer == NULL || flag_out == NULL ||
+      hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_STATE0_REG_OFFSET);
+
+  *flag_out = mmio_region_get_bit32(timer->base_addr, reg, comp_id);
+
+  return kDifRvTimerOk;
+}
+
+dif_rv_timer_result_t dif_rv_timer_irq_clear(const dif_rv_timer_t *timer,
+                                             uint32_t hart_id,
+                                             uint32_t comp_id) {
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  ptrdiff_t reg = irq_reg_for_hart(hart_id, timer->config.comparator_count,
+                                   RV_TIMER_INTR_STATE0_REG_OFFSET);
+
+  mmio_region_write32(timer->base_addr, reg, 1 << comp_id);
```

Comment:
bitfield? (applies to other usages as well).

Reply:
I was actually hoping to leave this alone until @lenary got done with his refactor of mmio/bitfield. The code I'd like to write here is something like
```c
mmio_region_write32(blah, blah, bitfield_set_bit32(0, comp_id, true));
```
or something like that.